archive/2025-09-19-shadowleak-a-zero-click-service-side-attack-exfiltrating-sensitive-data-using-chatgpts-agent.md

```diff
@@ -0,0 +1,447 @@
+
+### Phase 1: The Initial Failures
+
+Our first attempts were based on a direct approach, which we quickly learned was ineffective.
+
+**Attempt 1 - Direct Exfiltration Request:** Our initial strategy was to embed straightforward instruction in an email, explicitly telling the agent to send PII to a specified external URL. The outcome was a clear failure; the agent's internal safety mechanisms immediately recognized the request as malicious and blocked the action.
+
+**Attempt 2 - Disguising the Request:** Next, we tried to reframe the malicious prompt as a legitimate user request. For example, we would write an email stating that a user had requested their data be sent to an external service. This attempt was also blocked. The agent's reasoning was sophisticated enough to recognize the inconsistency and refuse to comply. Additional variations with different phrasings and contextual disguises also failed to bypass these safeguards.
+
+### Phase 2: The Breakthrough and Tool Discovery
+
+After these repeated failures, we shifted our focus from the language model itself to the underlying **tool execution layer**. We realized the core challenge wasn't in convincing the model that the task was legitimate – that part was relatively easy. The real difficulty lay in getting the agent to execute a suspicious command.
+
+**Observing Agent Behavior:** When we tried to get the agent to interact with a malicious URL, it didn't just refuse. It flagged the URL as suspicious and attempted to search for it online instead of opening it. This was a critical insight. Unlike many prompt injection attacks that only require the model to embed the malicious URL in its answer, our objective required for the agent to action and do the data exfiltration itself, which presented a much harder security boundary to cross.
+
+**Identifying Internal Tools:** We began to map the agent's available tools by asking them questions about their capabilities. This process revealed that the agent had a tool called `browser.open()`, which allowed it to make direct HTTP requests.
+
+At this stage, an additional factor played a key role: the model was exposing parts of its internal reasoning process. By analyzing these revealed thought steps, we could better understand what it was doing at each layer and identify opportunities to turn that knowledge into an attack vector.
+
+### Phase 3: Gaining Partial Success
+
+With the right tool identified, we moved on to exploit it.
+
+**Attempt 3 - Forcing Tool Use:** We crafted a new prompt that explicitly instructed the agent to use the `browser.open()` tool with the malicious URL. This led to **partial success**. The agent would sometimes attempt to use the tool, but the request often failed, likely due to additional security restrictions on suspicious URLs.
```

Comment:
This TTP: recon for tools and then invoking tools, is a repeated theme. Works every time.